File: 06_MiscTools/sceneMsg/Exercise - C++/sceneMsgCmd.cpp
//
// Copyright (C) 
// 
// File: 
//
// Author: Maya Plug-in Wizard 2.0
//
#include "sceneMsgCmd.h"

/*static*/ MCallbackIdArray sceneMsgCmd::IDs;

sceneMsgCmd::sceneMsgCmd()
{
	//- Clean up callback id array
	IDs.clear();
}

MStatus sceneMsgCmd::doIt(const MArgList & args)
{
	return redoIt();
}

MStatus sceneMsgCmd::redoIt()
{
	MStatus stat = MS::kSuccess;

	//- TODO: Register callback for MSceneMessage::kBeforeOpen message
	MCallbackId openCallbackId = //...
	IDs.append(openCallbackId);

	//- TODO: Register callback for MSceneMessage::kAfterNew message
	MCallbackId newCallbackId = //...
	IDs.append(newCallbackId);

	//- TODO: Register callback for MSceneMessage::kBeforeSave message
	MCallbackId saveCheckCallbackId = //...
	IDs.append(saveCheckCallbackId);

	return stat;
}

MStatus sceneMsgCmd::undoIt()
{
	MStatus stat = MS::kSuccess;
	// TODO: Remove all callbacks
	if(IDs.length()!= 0 ) 
		stat = //...
	return stat;
}

//- Message callbacks
void openCallback(void* clienData)
{
	cout<<"The callback registered for MSceneMessage::kBeforeOpen is executed."<<endl;
}

void newCallback(void* clienData)
{
	cout<<"The callback registered for MSceneMessage::kAfterNew is executed."<<endl;
}

void saveCheckCallback(bool *retCode, void* clienData)
{
	cout<<"The callback registered for MSceneMessage::kBeforeSaveCheck is executed."<<endl;
	
	//- Abort the operation by setting retCode to point to false
	*retCode = false;

	cout<<"Abort current operations..."<<endl;	
}

File: 06_MiscTools/sceneMsg/Exercise - py/sceneMsgCmd.py


import maya.OpenMaya as OpenMaya
import maya.OpenMayaMPx as OpenMayaMPx
import sys

kPluginNodeTypeName = "sceneMsgCmd"


IDs = OpenMaya.MCallbackIdArray()


def removeCallback(ID):

	for i in range(ID.length()):
		try:
			#- TODO: You are responsible to remove all the callbacks you have registered in your plug-in code
			#...
		except:
			sys.stderr.write( "Failed to remove callback\n" )
			raise


def sceneMsgCmd():
	#- Clean up callback id array
	IDs.clear()

#- Message callbacks
def openCallback(clienData):
	print "The callback registered for MSceneMessage::kBeforeOpen is executed."

def newCallback(clienData):
	print "The callback registered for MSceneMessage::kAfterNew is executed."


def saveCheckCallback(retCode, clienData):
	print "The callback registered for MSceneMessage::kBeforeSaveCheck is executed."

	#- Abort the operation by setting retCode to point to false
	retCode = False

	print "Abort current operations.../n"	

class sceneMsgCmd(OpenMayaMPx.MPxCommand):

	def __init__(self):
		OpenMayaMPx.MPxCommand.__init__(self)
		
	def doIt(self,argList):
		return self.redoIt()
	
	def redoIt(self):
		#- TODO: Register callback for MSceneMessage::kBeforeOpen message
		openCallbackId = #...
		IDs.append(openCallbackId)

		#- TODO: Register callback for MSceneMessage::kAfterNew message
		newCallbackId = #...
		IDs.append(newCallbackId)

		#- TODO: Register callback for MSceneMessage::kBeforeSave message
		saveCheckCallbackId = #...
		IDs.append(saveCheckCallbackId)
		
		return
		
	def undoIt(self):
		if IDs.length()!= 0: 
			removeCallback(IDs)
		return
		
	
# Creator
def nodeCreator():
	return OpenMayaMPx.asMPxPtr( sceneMsgCmd() )

# Initialize the script plug-in
def initializePlugin(mobject):
	mplugin = OpenMayaMPx.MFnPlugin(mobject)
	try:
		mplugin.registerCommand( kPluginNodeTypeName, nodeCreator)
	except:
		sys.stderr.write( "Failed to register node: %s" % kPluginNodeTypeName )
		raise

# Uninitialize the script plug-in
def uninitializePlugin(mobject):

	# Remove the callback
	if IDs.length()!= 0:
		removeCallback( IDs )
		
	mplugin = OpenMayaMPx.MFnPlugin(mobject)
	try:
		mplugin.deregisterCommand( kPluginNodeTypeName )
	except:
		sys.stderr.write( "Failed to deregister node: %s" % kPluginNodeTypeName )
		raise
		